proposals/p5434.md

```diff
@@ -0,0 +1,1331 @@
+  fn ValidLet[bound self: Self]() -> ref i32 {
+    return *self.pm;
+  }
+
+  // ✅ Valid
+  fn ValidVar[bound var self: Self]()
+      -> ref i32 { return *self.pm; }
+
+  var pm: i32*;
+}
+```
+
+The `bound` parameters must include everything referenced by the return, but
+need not be referenced, particularly not on every code path.
+
+```carbon
+// Result references `r` if `b` is true, and `p`
+// otherwise. Valid as long as both are marked `bound`.
+fn Conditional(b: bool, bound ref r: C, bound p: C*)
+    -> ref C {
+  if (b) {
+    return r;
+  } else {
+    return *p;
+  }
+}
+```
+
+The parameters of functions defined in an interface may also be marked as
+`bound`. The `impl` of that interface for a type can only have `bound` in a
+subset of the positions present in the interface.
+
+```carbon
+interface I {
+  fn F[bound ref self: Self]
+      (a: Self, bound b: Self, bound c: Self*)
+      -> ref Self;
+}
+
+impl C1 as I {
+  // ✅ Valid: matches interface
+  fn F[bound ref self: Self]
+      (a: Self, bound b: Self, bound c: Self*)
+      -> ref Self;
+}
+
+impl C2 as I {
+  // ✅ Valid: proper subset of `bound` params
+  fn F[ref self: Self]
+      (a: Self, bound b: Self, c: Self*)
+      -> ref Self;
+}
+
+impl C2 as I {
+  // ❌ Invalid: `a` is not bound in `I.F`.
+  fn F[ref self: Self]
+      (bound a: Self, b: Self, c: Self*)
+      -> ref Self;
+}
+```
+
+Like `[[clang::lifetimebound]]` in C++, `bound` does not affect semantics or
+calling conventions, just what code is legal. This helps avoid mismatches
+between typechecking against the signatures in an interface when the `impl`
+functions are different. Exception: the question of whether `bound` affects the
+lifetime of temporaries is [future work](#temporary-lifetimes).
+
+Note that all combinations of a `var`/`ref`/`let` return can be bound to a
+`var`/`ref`/`let` parameter. Examples:
+
+```carbon
+fn RefToLet(bound ref x: C) -> let D { return x.d; }
+fn LetToRef(bound y: C) -> ref D { return *y.ptr; }
+fn VarToRef(bound var p: i32*) -> ref i32 { return *p; }
+fn VarToVar(bound var p: i32*) -> var i32* { return p; }
+```
+
+For full safety, we need each bound parameter to be immutable for the duration
+of the lifetime of the returned result. However, the objective for now is only
+matching `[[clang::lifetimebound]]`, which has the goal of preventing some
+classes of bugs, not full memory safety. We will reconsider this with the memory
+safety design.
+
+### How addresses interact with `ref`
+
+The address of a `ref` binding is `noalias` and either `captures(none)` or
+`captures(ret: address, provenance)`, depending on whether the binding is marked
+`bound`.
+
+-   `noalias` means like C `restrict`; you can't observe mutations through
+    aliases; mutation through a restricted pointer is not observable through
+    another pointer
+-   `captures(none)` means no transitive escape, can pass a nocapture pointer to
+    another nocapture function, can't store to memory or return
```

Comment:
The omitted words in this sentence are hard to reconstruct. Is this right? If so I suggest making that change or something like it.

```suggestion
-   `captures(none)` means there is no transitive escape: you can pass a nocapture pointer to
    another nocapture function, but you can't store to memory or return
```